core/src/pow/types.rs

```diff
@@ -391,21 +391,54 @@ impl Proof {
 	}
 }
 
-#[inline(always)]
+fn extract_bits(bits: &Vec<u8>, bit_start: usize, bit_count: usize, read_from: usize) -> u64 {
+	let mut buf: [u8; 8] = [0; 8];
+	// read 8 bytes
+	buf.copy_from_slice(&bits[read_from..read_from + 8]);
+	// convert litl endian bits to u64
+	let mut nonce = u64::from_le_bytes(buf);
+	// shake off high bits (they are part of prev number)
+	nonce <<= 64 - (bit_start - read_from * 8) - bit_count;
+	// shake of low bits (next number if any)
+	nonce >> 64 - bit_count
+}
+
 fn read_number(bits: &Vec<u8>, bit_start: usize, bit_count: usize) -> u64 {
 	if bit_count == 0 {
 		return 0;
 	}
-	let mut buf: [u8; 8] = [0; 8];
-	let mut byte_start = bit_start / 8;
-	if byte_start + 8 > bits.len() {
-		byte_start = bits.len() - 8;
-	}
-	buf.copy_from_slice(&bits[byte_start..byte_start + 8]);
-	buf.reverse();
-	let mut nonce = u64::from_be_bytes(buf);
-	nonce = nonce << 64 - (bit_start - byte_start * 8) - bit_count;
-	nonce >> 64 - bit_count
+	// find where the first byte to read starts
+	let mut read_from = bit_start / 8;
+	// move back if we are too close to the end of bits
+	if read_from + 8 > bits.len() {
+		read_from = bits.len() - 8;
+	}
+	// calculate max bit we can read up to (+64 bits from the start)
+	let max_read_pos = (read_from + 8) * 8;
+	// calculate max bit we want to read
+	let max_pos = bit_start + bit_count;
+	// check if we can read it all at once
+	if max_read_pos >= max_pos {
+		extract_bits(bits, bit_start, bit_count, read_from)
+	} else {
+		// we need to read more than 8 bytes, we need to read twice
```

Comment:
how about simply reading one byte separately with

let low = bits[read_from];
let high =  extract_bits(bits, bit_start+8, bit_count, read_from+1);
(high << 8) + low;

Reply:
That avoid the need to compare read_from+8 against bits.len()

Reply:
very elegant, thanks